6541 dedup=[cksum],verify defeated spa feature check

ilovezfs opened the pull request:
zio_checksum_to_feature() expects a zio_checksum enum not a raw property
intval, so the new checksums weren't being detected when the
ZIO_CHECKSUM_VERIFY flag got in the way.

Given a pool without feature@sha512,

    zfs create -o dedup=sha512 naughty/fivetwelve_noverify_ds

would fail as expected since the raw intval would indeed be equal to
SPA_FEATURE_SHA512.

However,

    zfs create -o dedup=sha512,verify naughty/fivetwelve_verify_ds

would incorrectly succeed because ZIO_CHECKSUM_VERIFY would be in the
way, the raw intval would not be a member of the enum, and
zio_checksum_to_feature() would return SPA_FEATURE_NONE, with the result
that spa_feature_is_enabled() would never be called.

This was first detected with edonr, since in that case verify is
required.

This commit clears the ZIO_CHECKSUM_VERIFY flag before calling
zio_checksum_to_feature() and verifies in zio_checksum_to_feature() that
ZIO_CHECKSUM_VERIFY has been cleared to attempt to prevent the same bug
from occurring again in the future.

danmcd commented on usr/src/uts/common/fs/zfs/zio_checksum.c:
```diff
@@ -136,9 +136,13 @@ zio_checksum_info_t zio_checksum_table[ZIO_CHECKSUM_FUNCTIONS] = {
+ * The flag corresponding to the "verify" in dedup=[checksum,]verify
+ * must be cleared first.
+ */
+spa_feature_t zio_checksum_to_feature(enum zio_checksum cksum) {
```
cstyle nit.  Stick with the newline after spa_feature_t, please.

ilovezfs commented:
@danmcd Fixed. Thanks for spotting that.

danmcd commented on usr/src/uts/common/fs/zfs/zio_checksum.c:
```diff
@@ -136,9 +136,14 @@ zio_checksum_info_t zio_checksum_table[ZIO_CHECKSUM_FUNCTIONS] = {
 spa_feature_t
-zio_checksum_to_feature(enum zio_checksum cksum)
-{
+zio_checksum_to_feature(enum zio_checksum cksum) {
```
Ummm... sorry, move the '{' down too, please.  (Thanks for the other fix.)

ilovezfs replied:
@danmcd Done. Oh the hazards of vi gq{gq}.

ahrens commented:
@zettabot go

ahrens commented:
LGTM

ahrens commented:
Please check the tests, looks like a lint error:
"../../common/fs/zfs/zio_checksum.c", line 146: warning: precedence confusion possible; parenthesize (E_PRECEDENCE_CONFUSION)

ilovezfs commented:
@ahrens Fixed. BTW, would it be preferable to allow for less precise inputs, change the argument to int, and automatically clear the ZIO_CHECKSUM_VERIFY?

ahrens commented:
@zettabot go

ahrens commented:
@ilovezfs I assume you mean the argument to zio_checksum_to_feature().  No, I think the way you've done it is good.

ilovezfs commented:
@ahrens Yup, that's what I meant. OK, great.

ahrens commented:
@zettabot go

prakashsurya commented:
@zettabot go

ilovezfs commented:
Additional investigation shows that this does indeed cause panic (http://i.imgur.com/D0t7Okz.png) on OmniOS (Bloody release, r151017, omnios-2e8c0ba, OmniOS_Text_bloody_20151211.iso), as soon as you actually start to add some data to the dataset. This is the same panic observed on OS X, where it happens all by itself due to Spotlight, fsevents, etc. automatically adding some data as soon as the file system mounts.
```
zpool create -d planned c2t1d0
zfs create -o dedup=edonr,verify planned/demise
echo good > /panned/demise/bye.txt
... panic
```

Stack:
```
panic: assertition failed: feature_get_refcount(spa, feature, &refcount) != 48 (0x30 != 0x30), file: ../../common/fs/zfs/zfeature.c, line: 390
zfs:feature_do_action
zfs:spa_feature_incr
zfs:dsl_dataset_activate_feature
zfs:dsl_dataset_sync
zfs:dsl_pool_sync
zfs:spa_sync
zfs:txg_sync_thread
unix:thread_start
````

prakashsurya commented:
It looks like this failed the regression tests in the recent runs:

1. http://build.prakashsurya.com:8080/job/openzfs-regression-tests/138/
2. http://build.prakashsurya.com:8080/job/openzfs-regression-tests/140/

To help with debugging these failures, I've tar-ed up the workspace of the ztest failures, configured anonymous FTP, and put the tarballs up for download. Assuming I configured things correctly, one should be able to use any FTP client to download the files, e.g. I can point my chrome browser at `ftp://build.prakashsurya.com`. There's two files:

1. ftp://build.prakashsurya.com/openzfs-run-ztest-138.tar.gz - This contains the ztest results for the first run, openzfs-regression-tests 138
2. ftp://build.prakashsurya.com/openzfs-run-ztest-138.tar.gz - This contains the ztest results for the second run, openzfs-regression-tests 140

There's a core file in the second run from the failed ztest, but the first run didn't tigger crash ztest so there's no core file. An illumos system might be necessary (i.e. `mdb`) to properly analyze the core file, but my guess is any implementation of `zdb` (e.g. illumos, linux, mac, freebsd, w/e) should be able to be used to inspect the ztest vdev files contained in the tarballs.

ilovezfs commented:
@prakashsurya Worked. Thanks.

@ahrens I'm wondering if it would be useful to re-run the tests without the VERIFY at
https://github.com/openzfs/openzfs/pull/51/files#diff-f8f4f6d3ef256a7930d4f36b0052e4f0R146

ahrens commented:
@ilovezfs I'm happy to tell zettabot to run if you update the pull request, but how would that be useful?  Are we hitting that VERIFY?

ilovezfs commented:
@ahrens I don't know if we are, and I think it's unlikely. But since master is clearly wrong and can cause easily reproducible panic on every platform with a single zfs create in a pool without the new feature flag flipped on, I'm at a loss for what exactly would cause the tests to fail IF they are actually known to pass elsewhere in equivalent environments. I'm thinking it's quite likely that either a) the environment in which the tests for this PR run is somehow different from the known passing environments, or b) the environments for other PRs are indeed equivalent and also fail (with that failure possibly unnoticed or undetected?).

So I'm wishfully hoping it could be as simple as something unexpectedly hitting that verify. Of course, the "scariest" explanation would be if somehow this PR is revealing a bug with the new dedup checksums themselves, though I think that, too, is unlikely.

In any case, illumos HEAD is basically unreleasable right now due to this bug since <code>zfs create -o dedup=S,verify tank/foo</code> is enough to send any system into an "endless" panic-reboot-panic-reboot loop if the system has any preexisting pool tank with an S in {edonr, sha512, skein} such that feature@S has not been explicitly enabled on tank. The system and pool are, of course, fixable once someone figures out what went wrong, but that's beside the point.

ilovezfs commented:
@ahrens I commented out the VERIFY if you'd like to try it without that.

ahrens commented:
@zettabot go

ilovezfs commented:
@ahrens hmm...so it passed now? Should we try again with the VERIFY put back in? :)

ahrens commented:
sure

ilovezfs commented:
@ahrens OK, back the way it was.

ahrens commented:
@zettabot go

ahrens commented:
@ilovezfs These changes look ready to me.  How have you tested them (besides the regression tests run here)?  E.g. are these changes present in another platform, and/or have you done manual testing to check that "dedup=sha512,verify" errors out when the feature is disabled?

ilovezfs commented:
@ahrens Yes, I've merged it to OpenZFS on OS X, and it does prevent creation of a dataset with dedup=sha512,verify, etc. unless the corresponding feature is enabled. However, I just retested, and noticed that the machine still panics after mount and writing some data with this stack:

```
Mon Jan 25 16:52:23 2016
panic(cpu 0 caller 0xffffff7fade6e1ee): "VERIFY3(" "feature_get_refcount(spa, feature, &refcount)" " " "!=" " " "45" ") " "failed (" "45" " " "!=" " " "45" ")\n"@zfeature.c:395
Backtrace (CPU 0), Frame : Return Address
0xffffff820421bbd0 : 0xffffff802b61d6d6 mach_kernel : _panic + 0xc6
0xffffff820421bc40 : 0xffffff7fade6e1ee net.lundman.zfs : _feature_do_action + 0x56
0xffffff820421bc80 : 0xffffff7fade18892 net.lundman.zfs : _dmu_object_zapify + 0xc9
0xffffff820421bcb0 : 0xffffff7fade29ed1 net.lundman.zfs : _dsl_dataset_activate_feature + 0x8a
0xffffff820421bd10 : 0xffffff7fade2b59f net.lundman.zfs : _dsl_dataset_sync + 0x251
0xffffff820421bd50 : 0xffffff7fade372fa net.lundman.zfs : _dsl_pool_sync + 0x13c
0xffffff820421bdd0 : 0xffffff7fade53512 net.lundman.zfs : _spa_sync + 0x3b1
0xffffff820421bed0 : 0xffffff7fade5b85e net.lundman.zfs : _txg_sync_thread + 0x454
0xffffff820421bfb0 : 0xffffff802b6b34b7 mach_kernel : _call_continuation + 0x17
      Kernel Extensions in backtrace:
         net.lundman.zfs(1.5)[AC464F66-CE5A-3042-B704-089B1AB887FA]@0xffffff7fade06000->0xffffff7fae044fff
            dependency: com.apple.iokit.IOStorageFamily(1.8)[A3CC4E44-8E10-3D9A-BA8E-95743E79D125]@0xffffff7fabb79000
            dependency: net.lundman.spl(1.5.0)[218BEB87-05BF-3684-BE1A-74B6A51DDB29]@0xffffff7faddb8000
```

So clearly it's not fully correct yet.

ilovezfs commented:
@ahrens I think the problem is that ZIO_CHECKSUM_VERIFY has to be cleared here, too:
https://github.com/illumos/illumos-gate/blob/master/usr/src/uts/common/fs/zfs/dsl_dataset.c#L137

Even though ZIO_CHECKSUM_VERIFY is not actually a member of the zio_checksum enum, it can still be in the io_bp due to
https://github.com/illumos/illumos-gate/blob/master/usr/src/uts/common/fs/zfs/dmu_objset.c#L197

ilovezfs commented:
Nope, still panics with
```
diff --git a/module/zfs/dsl_dataset.c b/module/zfs/dsl_dataset.c
index d647b46..dd229b8 100644
--- a/module/zfs/dsl_dataset.c
+++ b/module/zfs/dsl_dataset.c
@@ -137,7 +137,7 @@ dsl_dataset_block_born(dsl_dataset_t *ds, const blkptr_t *bp, dmu_tx_t *tx)
                        B_TRUE;
        }
 
-       spa_feature_t f = zio_checksum_to_feature(BP_GET_CHECKSUM(bp));
+       spa_feature_t f = zio_checksum_to_feature(BP_GET_CHECKSUM(bp) & ~ZIO_CHECKSUM_VERIFY);
        if (f != SPA_FEATURE_NONE)
                ds->ds_feature_activation_needed[f] = B_TRUE;
```
So I'll need to investigate further.

ilovezfs commented:
@ahrens OK, I've fixed the panic, but I'm not sure it's the correct fix. It seems that the sha512, edonr, and skein features depend on SPA_FEATURE_EXTENSIBLE_DATASET but they're not registered with that as a dependency. Was it intended that they should depend on extensible dataset, or is that itself a bug? I've added extensible dataset as a dependency for each of them in a separate commit in this same pull request. Also, assuming the dependency is intended, I'm not sure if you want a separate dependency array for each of the new checksum features or a single shared one, since they're all the same.
https://github.com/ilovezfs/openzfs/commit/5a3f98b338c7c21586ba4c16e7357e5ed0460b58

Also, I was incorrect about the io_bp having ZIO_CHECKSUM_VERIFY. The ZIO_CHECKSUM_MASK is applied before the io_bp's checksum value is set. So I've updated the original commit in this pull request to use ZIO_CHECKSUM_MASK instead of ~ZIO_CHECKSUM_VERIFY, since that seems to be the preexisting way of clearing the verify flag. Do you like that change, or would you prefer I revert it back to using ~ZIO_CHECKSUM_VERIFY?
https://github.com/ilovezfs/openzfs/commit/1a832d2a2fe3a187b84f021fc333cf5ba6c974ce

Last, should the new-checksums-extensible-dataset bug get its own issue in the illumos issue tracker or piggy back on 6541?

ilovezfs commented:
@rlaager The fixes for the man page and whitespace problems you flagged, as well as for a couple other whitespace anomalies I noticed, are now in there, if you'd like to look it over.

@mahrens Any thoughts on those questions above, and is this looking ready for Mr. Zetta Bot's water tortures?

ahrens commented:
@zettabot go

rlaager commented on usr/src/common/zfs/zfeature_common.c:
```diff
@@ -210,8 +212,8 @@ zpool_feature_init(void)
 	static const spa_feature_t filesystem_limits_deps[] = {
-	    SPA_FEATURE_EXTENSIBLE_DATASET,
-	    SPA_FEATURE_NONE
+		SPA_FEATURE_EXTENSIBLE_DATASET,
```
Whitespace issue here still.

ilovezfs replied:
Do you see the issue in the raw version of the file?

rlaager replied:
Sorry, my bad. That one is correct. Instead, skein_deps & edonr_deps are wrong, as they are all spaces (no tabs) right now.

ilovezfs commented:
```
ztest: '/usr/sbin/i86/zdb -bccsv -d -U /var/tmp/jenkins/workspace/openzfs-run-ztest/zpool.cache ztest' exit code 2
running: echo ::stack
mdb: failed to open core: No such file or directory
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```
Not sure what, if anything, to make of that.

ilovezfs commented:
@rlaager How's that?

@ahrens I created the two additional illumos issues and prepended their numbers to the commit titles.

rlaager commented:
It's fixed now, but it was broken last night. I thought I was losing my mind, but ZFS snapshots on my local system prove there was a difference:

rlaager@watermelon:~/.zfs/snapshot/auto-hourly-2016-01-28-2200$ md5sum zfeature_common.c ~/openzfs/usr/src/common/zfs/zfeature_common.c
497922fa43e10878baca74a217eaf7f3  zfeature_common.c
fe44104c931430605cd04fef86723847  /home/rlaager/openzfs/usr/src/common/zfs/zfeature_common.c
rlaager@watermelon:~/.zfs/snapshot/auto-hourly-2016-01-28-2200$ diff -u zfeature_common.c ~/openzfs/usr/src/common/zfs/zfeature_common.c
--- zfeature_common.c	2016-01-28 21:27:25.792291064 -0600
+++ /home/rlaager/openzfs/usr/src/common/zfs/zfeature_common.c	2016-01-29 15:56:50.643438415 -0600
@@ -238,25 +238,25 @@
 	static const spa_feature_t sha512_deps[] = {
 		SPA_FEATURE_EXTENSIBLE_DATASET,
 		SPA_FEATURE_NONE
-        };
+	};
 	zfeature_register(SPA_FEATURE_SHA512,
 	    "org.illumos:sha512", "sha512",
 	    "SHA-512/256 hash algorithm.",
 	    ZFEATURE_FLAG_PER_DATASET, sha512_deps);
 
-        static const spa_feature_t skein_deps[] = {
-                SPA_FEATURE_EXTENSIBLE_DATASET,
-                SPA_FEATURE_NONE
-        };
+	static const spa_feature_t skein_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
+		SPA_FEATURE_NONE
+	};
 	zfeature_register(SPA_FEATURE_SKEIN,
 	    "org.illumos:skein", "skein",
 	    "Skein hash algorithm.",
 	    ZFEATURE_FLAG_PER_DATASET, skein_deps);
 
-        static const spa_feature_t edonr_deps[] = {
-                SPA_FEATURE_EXTENSIBLE_DATASET,
-                SPA_FEATURE_NONE
-        };
+	static const spa_feature_t edonr_deps[] = {
+		SPA_FEATURE_EXTENSIBLE_DATASET,
+		SPA_FEATURE_NONE
+	};
 	zfeature_register(SPA_FEATURE_EDONR,
 	    "org.illumos:edonr", "edonr",
 	    "Edon-R hash algorithm.",

ilovezfs commented:
@rlaager Yeah, thank you for catching that. I'm wondering if we should move all of the dependency arrays up to the top of the function so that the whitespace is all tab-indented elements in one section and all four-space continuations in the other. Visually it may look better and prevent the issue in the future by making the correct spacing obvious and allowing for copy-paste of the array or registration directly above. Also, it would make it more visually obvious which features have dependencies and which do not. On the other hand, it separates the registration from the array, so some might find that less convenient to have to cross reference.

ahrens commented:
@zettabot go

ahrens commented:
@ilovezfs You're right, the new checksum algo features should depend on SPA_FEATURE_EXTENSIBLE_DATASET, because they use ZFEATURE_FLAG_PER_DATASET.  It would be nice if zfeature_register() could verify that if ZFEATURE_FLAG_PER_DATASET is set, then the feature depends on SPA_FEATURE_EXTENSIBLE_DATASET.

Declaring a different dependency array per feature, as you've done, is fine.

ZIO_CHECKSUM_MASK seems like the right way to go.

Changes look good to me.

ahrens commented:
@rlaager Can we count you as a code reviewer?

ilovezfs commented:
@ahrens 
> It would be nice if zfeature_register() could verify that if ZFEATURE_FLAG_PER_DATASET is set, then the feature depends on SPA_FEATURE_EXTENSIBLE_DATASET

Is the converse supposed to be true?
I noticed bookmarks and filesystem_limits have SPA_FEATURE_EXTENSIBLE_DATASET but are ZFEATURE_FLAG_READONLY_COMPAT, not ZFEATURE_FLAG_PER_DATASET or (ZFEATURE_FLAG_READONLY_COMPAT | ZFEATURE_FLAG_PER_DATASET).

ahrens commented:
> Is the converse supposed to be true?

No.

> I noticed bookmarks and filesystem_limits have SPA_FEATURE_EXTENSIBLE_DATASET but are ZFEATURE_FLAG_READONLY_COMPAT, not ZFEATURE_FLAG_PER_DATASET or (ZFEATURE_FLAG_READONLY_COMPAT | ZFEATURE_FLAG_PER_DATASET).

That sounds correct to me.

ilovezfs commented:
@ahrens OK, thanks. Then how's this?

```
diff --git a/module/zfs/zfeature_common.c b/module/zfs/zfeature_common.c
index b11e647..fe20482 100644
--- a/module/zfs/zfeature_common.c
+++ b/module/zfs/zfeature_common.c
@@ -134,6 +134,19 @@ zfeature_depends_on(spa_feature_t fid, spa_feature_t check) {
 	return (B_FALSE);
 }
 
+static boolean_t
+deps_contains_feature(const spa_feature_t *deps, const spa_feature_t feature)
+{
+	spa_feature_t *p = (spa_feature_t *)deps;
+
+	while (p && (*p != SPA_FEATURE_NONE)) {
+		if (*p == feature)
+			return (B_TRUE);
+		p++;
+        }
+        return (B_FALSE);
+}
+
 static void
 zfeature_register(spa_feature_t fid, const char *guid, const char *name,
     const char *desc, zfeature_flags_t flags, const spa_feature_t *deps)
@@ -148,6 +161,9 @@ zfeature_register(spa_feature_t fid, const char *guid, const char *name,
 	ASSERT3U(fid, <, SPA_FEATURES);
 	ASSERT(zfeature_is_valid_guid(guid));
 
+	VERIFY(((flags & ZFEATURE_FLAG_PER_DATASET) == 0) ||
+	    (deps_contains_feature(deps, SPA_FEATURE_EXTENSIBLE_DATASET)));
+
 	if (deps == NULL)
 		deps = nodeps;
 

```

ahrens commented:
@ilovezfs That looks good.  Not sure why you need to cast away the const.  Or check for p being null.  You might use a for loop rather than while:

`for (int i = 0; deps[i] != SPA_FEATURE_NONE; i++)`

rlaager commented:
@ahrens Yes, I'm comfortable being counted as a code reviewer. I re-reviewed everything again just now. I think I've reviewed it as well as I can, given my personal level of knowledge of the ZFS codebase.

ilovezfs commented:
@ahrens How's this https://github.com/ilovezfs/openzfs/commit/442c08b2dc2a6e26d1be72be57fa90bc7c196e7d

And would you like it added to this PR or a new one?

ahrens commented:
@ilovezfs that check looks good to me.

ahrens commented:
@ilovezfs And I'm fine with it being part of this PR.

ilovezfs commented:
@ahrens OK, great. And it's now in this PR.

ahrens commented:
@zettabot go

ilovezfs commented:
Woo-hoo, it passed! But probably best not to make eye contact with it, lest it get unhappy again.
